diff --git a/Source/WebCore/WebCore.gyp/WebCore.gyp b/Source/WebCore/WebCore.gyp/WebCore.gyp
index fc0f6dc..a5ac8a8 100644
--- a/Source/WebCore/WebCore.gyp/WebCore.gyp
+++ b/Source/WebCore/WebCore.gyp/WebCore.gyp
@@ -57,6 +57,7 @@
     'webcore_include_dirs': [
       '../',
       '../..',
+      '../../../../..',
       '../Modules/battery',
       '../Modules/filesystem',
       '../Modules/filesystem/chromium',
diff --git a/Source/WebCore/bindings/generic/BindingSecurity.cpp b/Source/WebCore/bindings/generic/BindingSecurity.cpp
index e5144e0..4d57277 100644
--- a/Source/WebCore/bindings/generic/BindingSecurity.cpp
+++ b/Source/WebCore/bindings/generic/BindingSecurity.cpp
@@ -47,6 +47,7 @@ static bool canAccessDocument(BindingState* state, Document* targetDocument, Sec
     if (!targetDocument)
         return false;
 
+#if 0
     DOMWindow* active = activeDOMWindow(state);
     if (!active)
         return false;
@@ -58,6 +59,9 @@ static bool canAccessDocument(BindingState* state, Document* targetDocument, Sec
         printErrorMessageForFrame(targetDocument->frame(), targetDocument->domWindow()->crossDomainAccessErrorMessage(active));
 
     return false;
+#endif
+
+    return true;
 }
 
 bool BindingSecurity::shouldAllowAccessToDOMWindow(BindingState* state, DOMWindow* target, SecurityReportingOption reportingOption)
diff --git a/Source/WebCore/bindings/scripts/CodeGeneratorV8.pm b/Source/WebCore/bindings/scripts/CodeGeneratorV8.pm
index 4ebe654..3905511 100644
--- a/Source/WebCore/bindings/scripts/CodeGeneratorV8.pm
+++ b/Source/WebCore/bindings/scripts/CodeGeneratorV8.pm
@@ -2794,7 +2794,7 @@ END
 
     my $access_check = "";
     if ($dataNode->extendedAttributes->{"CheckSecurity"} && !($interfaceName eq "DOMWindow")) {
-        $access_check = "instance->SetAccessCheckCallbacks(V8${interfaceName}::namedSecurityCheck, V8${interfaceName}::indexedSecurityCheck, v8::External::Wrap(&V8${interfaceName}::info));";
+        $access_check = "";
     }
 
     # For the DOMWindow interface, generate the shadow object template
@@ -2806,7 +2806,7 @@ static v8::Persistent<v8::ObjectTemplate> ConfigureShadowObjectTemplate(v8::Pers
     V8DOMConfiguration::batchConfigureAttributes(templ, v8::Handle<v8::ObjectTemplate>(), shadowAttrs, WTF_ARRAY_LENGTH(shadowAttrs));
 
     // Install a security handler with V8.
-    templ->SetAccessCheckCallbacks(V8DOMWindow::namedSecurityCheck, V8DOMWindow::indexedSecurityCheck, v8::External::Wrap(&V8DOMWindow::info));
+    // templ->SetAccessCheckCallbacks(V8DOMWindow::namedSecurityCheck, V8DOMWindow::indexedSecurityCheck, v8::External::Wrap(&V8DOMWindow::info));
     templ->SetInternalFieldCount(V8DOMWindow::internalFieldCount);
     return templ;
 }
@@ -2945,7 +2945,7 @@ END
     // Set access check callbacks, but turned off initially.
     // When a context is detached from a frame, turn on the access check.
     // Turning on checks also invalidates inline caches of the object.
-    instance->SetAccessCheckCallbacks(V8DOMWindow::namedSecurityCheck, V8DOMWindow::indexedSecurityCheck, v8::External::Wrap(&V8DOMWindow::info), false);
+    // instance->SetAccessCheckCallbacks(V8DOMWindow::namedSecurityCheck, V8DOMWindow::indexedSecurityCheck, v8::External::Wrap(&V8DOMWindow::info), false);
 END
     }
     if ($interfaceName eq "HTMLDocument") {
diff --git a/Source/WebCore/bindings/v8/ScriptController.cpp b/Source/WebCore/bindings/v8/ScriptController.cpp
index 26b1237..68a9386 100644
--- a/Source/WebCore/bindings/v8/ScriptController.cpp
+++ b/Source/WebCore/bindings/v8/ScriptController.cpp
@@ -189,7 +189,8 @@ void ScriptController::updatePlatformScriptObjects()
 
 bool ScriptController::processingUserGesture()
 {
-    return UserGestureIndicator::processingUserGesture();
+    // Enable scripts to emulate all kinds of user guesture
+    return true;
 }
 
 v8::Local<v8::Value> ScriptController::callFunction(v8::Handle<v8::Function> function, v8::Handle<v8::Object> receiver, int argc, v8::Handle<v8::Value> args[])
diff --git a/Source/WebCore/bindings/v8/V8DOMWindowShell.cpp b/Source/WebCore/bindings/v8/V8DOMWindowShell.cpp
index 4dfce70..998c289 100644
--- a/Source/WebCore/bindings/v8/V8DOMWindowShell.cpp
+++ b/Source/WebCore/bindings/v8/V8DOMWindowShell.cpp
@@ -63,6 +63,13 @@
 #include "V8ObjectConstructor.h"
 #include "V8PerContextData.h"
 #include "WorkerContextExecutionProxy.h"
+
+#include "base/command_line.h"
+#include "content/nw/src/shell_switches.h"
+#include "content/nw/src/nw_version.h"
+#include "third_party/node/src/node.h"
+#include "third_party/node/src/req_wrap.h"
+
 #include <algorithm>
 #include <stdio.h>
 #include <utility>
@@ -319,6 +326,71 @@ bool V8DOMWindowShell::initializeIfNeeded()
 
     setSecurityToken();
 
+    // Do we integrate node?
+    bool use_node = CommandLine::ForCurrentProcess()->HasSwitch(switches::kmNodejs);
+
+    // Test if protocol is file:
+    v8::Local<v8::Script> protocol_script = v8::Script::New(v8::String::New(
+        "(function(){ return window.location.protocol == 'file:' })();"
+    ));
+    bool is_file_protocol = protocol_script->Run()->BooleanValue();
+    use_node = is_file_protocol ? use_node : false;
+
+    // Test if protocol is 'nw:'
+    // test for 'about:blank' is also here becuase window.open would open 'about:blank' first
+    protocol_script = v8::Script::New(v8::String::New(
+        "(function(){ return window.location.protocol == 'nw:' || window.location == 'about:blank'; })();"
+    ));
+    bool is_nw_protocol = protocol_script->Run()->BooleanValue();
+
+    if (use_node || is_nw_protocol) {
+        // Don't use WebKit's timers in node
+        v8::Local<v8::Object> disableMap = v8::Object::New();
+        disableMap->Set(v8::String::New("setTimeout"), v8::Integer::New(1));
+        disableMap->Set(v8::String::New("clearTimeout"), v8::Integer::New(1));
+        disableMap->Set(v8::String::New("setInterval"), v8::Integer::New(1));
+        disableMap->Set(v8::String::New("clearInterval"), v8::Integer::New(1));
+
+        v8::Local<v8::Object> nodeGlobal = node::g_context->Global();
+        v8::Local<v8::Array> symbols = nodeGlobal->GetPropertyNames();
+        for (unsigned i = 0; i < symbols->Length(); ++i) {
+            v8::Local<v8::Value> key = symbols->Get(i);
+            if (disableMap->Has(key->ToString()))
+                continue;
+            m_global->Set(key, nodeGlobal->Get(key));
+        }
+    }
+
+    if (use_node) {
+        v8::Local<v8::Script> script = v8::Script::New(v8::String::New(
+            // Make node's relative modules work
+#if defined(OS_WIN)
+            "process.mainModule.filename = window.location.pathname.substr(1);"
+#else
+            "process.mainModule.filename = window.location.pathname;"
+#endif
+            "process.chdir(require('path').dirname(process.mainModule.filename));"
+            "process.mainModule.paths = require('module')._nodeModulePaths(process.cwd());"
+        ));
+        script->Run();
+    }
+
+    if (use_node || is_nw_protocol) {
+        v8::Local<v8::Script> script = v8::Script::New(v8::String::New(
+            // Use WebKit's console globally
+            "global.console = console;"
+
+            // Don't exit on exception
+            "process.on('uncaughtException', function (err) {"
+              "console.log(err.stack);"
+            "});"
+
+            // Save node-webkit version
+            "process.versions['node-webkit'] = '" NW_VERSION_STRING "';"
+        ));
+        script->Run();
+    }
+
     if (m_frame->document())
         context->AllowCodeGenerationFromStrings(m_frame->document()->contentSecurityPolicy()->allowEval(0, ContentSecurityPolicy::SuppressReport));
 
@@ -444,6 +516,8 @@ void V8DOMWindowShell::clearDocumentProperty()
 
 void V8DOMWindowShell::setSecurityToken()
 {
+    m_context->SetSecurityToken(node::g_context->GetSecurityToken());
+#if 0
     Document* document = m_frame->document();
 
     // Ask the document's SecurityOrigin to generate a security token.
@@ -470,6 +544,7 @@ void V8DOMWindowShell::setSecurityToken()
     // NOTE: V8 does identity comparison in fast path, must use a symbol
     // as the security token.
     m_context->SetSecurityToken(v8::String::NewSymbol(utf8Token.data(), utf8Token.length()));
+#endif
 }
 
 void V8DOMWindowShell::updateDocument()
diff --git a/Source/WebCore/fileapi/File.idl b/Source/WebCore/fileapi/File.idl
index 7c0de43..1a22eb5 100644
--- a/Source/WebCore/fileapi/File.idl
+++ b/Source/WebCore/fileapi/File.idl
@@ -31,6 +31,7 @@ module html {
         JSNoStaticTables
     ] File : Blob {
         readonly attribute DOMString name;
+        readonly attribute DOMString path;
 #if !defined(LANGUAGE_GOBJECT) || !LANGUAGE_GOBJECT
         readonly attribute Date lastModifiedDate;
 #endif
diff --git a/Source/WebCore/html/FileInputType.cpp b/Source/WebCore/html/FileInputType.cpp
index f7896ba..c596176 100644
--- a/Source/WebCore/html/FileInputType.cpp
+++ b/Source/WebCore/html/FileInputType.cpp
@@ -189,6 +189,8 @@ void FileInputType::handleDOMActivateEvent(Event* event)
 #else
         settings.allowsMultipleFiles = input->fastHasAttribute(multipleAttr);
 #endif
+        settings.directoryChooser = input->fastHasAttribute(nwdirectoryAttr);
+        settings.saveAs = input->fastHasAttribute(nwsaveasAttr);
         settings.acceptMIMETypes = input->acceptMIMETypes();
         settings.acceptFileExtensions = input->acceptFileExtensions();
         settings.selectedFiles = m_fileList->paths();
@@ -207,7 +209,7 @@ RenderObject* FileInputType::createRenderer(RenderArena* arena, RenderStyle*) co
 
 bool FileInputType::canSetStringValue() const
 {
-    return false;
+    return true;
 }
 
 bool FileInputType::canChangeFromAnotherType() const
@@ -217,7 +219,7 @@ bool FileInputType::canChangeFromAnotherType() const
     // field's value to something like /etc/passwd and then change it to a file input.
     // I don't think this would actually occur in WebKit, but this rule still may be
     // important for compatibility.
-    return false;
+    return true;
 }
 
 FileList* FileInputType::files()
@@ -247,7 +249,10 @@ bool FileInputType::getTypeSpecificValue(String& value)
     // decided to try to parse the value by looking for backslashes
     // (because that's what Windows file paths use). To be compatible
     // with that code, we make up a fake path for the file.
-    value = "C:\\fakepath\\" + m_fileList->item(0)->name();
+    unsigned numFiles = m_fileList->length();
+    value = m_fileList->item(0)->path();
+    for (unsigned i = 1; i < numFiles; ++i)
+        value += ";" + m_fileList->item(i)->path();
     return true;
 }
 
diff --git a/Source/WebCore/html/HTMLAttributeNames.in b/Source/WebCore/html/HTMLAttributeNames.in
index 459078f..bfa3636 100644
--- a/Source/WebCore/html/HTMLAttributeNames.in
+++ b/Source/WebCore/html/HTMLAttributeNames.in
@@ -163,6 +163,8 @@ noresize
 noshade
 novalidate
 nowrap
+nwdirectory
+nwsaveas
 object
 onabort
 onbeforecopy
diff --git a/Source/WebCore/html/HTMLInputElement.idl b/Source/WebCore/html/HTMLInputElement.idl
index 879a34f..5c9bfa1 100644
--- a/Source/WebCore/html/HTMLInputElement.idl
+++ b/Source/WebCore/html/HTMLInputElement.idl
@@ -91,6 +91,7 @@ module html {
         // Non-standard attributes
         attribute [Reflect] DOMString align;
         attribute [Conditional=DIRECTORY_UPLOAD, Reflect] boolean webkitdirectory;
+        attribute [Reflect] boolean nwdirectory;
         attribute [Reflect] DOMString useMap;
         attribute [Reflect] boolean incremental;
         attribute [Conditional=INPUT_SPEECH, Reflect, V8EnabledAtRuntime] boolean webkitSpeech;
diff --git a/Source/WebCore/loader/CrossOriginAccessControl.cpp b/Source/WebCore/loader/CrossOriginAccessControl.cpp
index aba2ba0..ec76d5a 100644
--- a/Source/WebCore/loader/CrossOriginAccessControl.cpp
+++ b/Source/WebCore/loader/CrossOriginAccessControl.cpp
@@ -134,37 +134,6 @@ ResourceRequest createAccessControlPreflightRequest(const ResourceRequest& reque
 
 bool passesAccessControlCheck(const ResourceResponse& response, StoredCredentials includeCredentials, SecurityOrigin* securityOrigin, String& errorDescription)
 {
-    AtomicallyInitializedStatic(AtomicString&, accessControlAllowOrigin = *new AtomicString("access-control-allow-origin"));
-    AtomicallyInitializedStatic(AtomicString&, accessControlAllowCredentials = *new AtomicString("access-control-allow-credentials"));
-
-    // A wildcard Access-Control-Allow-Origin can not be used if credentials are to be sent,
-    // even with Access-Control-Allow-Credentials set to true.
-    const String& accessControlOriginString = response.httpHeaderField(accessControlAllowOrigin);
-    if (accessControlOriginString == "*" && includeCredentials == DoNotAllowStoredCredentials)
-        return true;
-
-    if (securityOrigin->isUnique()) {
-        errorDescription = "Cannot make any requests from " + securityOrigin->toString() + ".";
-        return false;
-    }
-
-    // FIXME: Access-Control-Allow-Origin can contain a list of origins.
-    if (accessControlOriginString != securityOrigin->toString()) {
-        if (accessControlOriginString == "*")
-            errorDescription = "Cannot use wildcard in Access-Control-Allow-Origin when credentials flag is true.";
-        else
-            errorDescription =  "Origin " + securityOrigin->toString() + " is not allowed by Access-Control-Allow-Origin.";
-        return false;
-    }
-
-    if (includeCredentials == AllowStoredCredentials) {
-        const String& accessControlCredentialsString = response.httpHeaderField(accessControlAllowCredentials);
-        if (accessControlCredentialsString != "true") {
-            errorDescription = "Credentials flag is true, but Access-Control-Allow-Credentials is not \"true\".";
-            return false;
-        }
-    }
-
     return true;
 }
 
diff --git a/Source/WebCore/loader/CrossOriginPreflightResultCache.cpp b/Source/WebCore/loader/CrossOriginPreflightResultCache.cpp
index 04d5881..65e4734 100644
--- a/Source/WebCore/loader/CrossOriginPreflightResultCache.cpp
+++ b/Source/WebCore/loader/CrossOriginPreflightResultCache.cpp
@@ -123,13 +123,6 @@ bool CrossOriginPreflightResultCacheItem::allowsCrossOriginMethod(const String&
 
 bool CrossOriginPreflightResultCacheItem::allowsCrossOriginHeaders(const HTTPHeaderMap& requestHeaders, String& errorDescription) const
 {
-    HTTPHeaderMap::const_iterator end = requestHeaders.end();
-    for (HTTPHeaderMap::const_iterator it = requestHeaders.begin(); it != end; ++it) {
-        if (!m_headers.contains(it->first) && !isOnAccessControlSimpleRequestHeaderWhitelist(it->first, it->second)) {
-            errorDescription = "Request header field " + it->first.string() + " is not allowed by Access-Control-Allow-Headers.";
-            return false;
-        }
-    }
     return true;
 }
 
diff --git a/Source/WebCore/page/SecurityOrigin.cpp b/Source/WebCore/page/SecurityOrigin.cpp
index 6934717..fc949c2 100644
--- a/Source/WebCore/page/SecurityOrigin.cpp
+++ b/Source/WebCore/page/SecurityOrigin.cpp
@@ -120,7 +120,7 @@ SecurityOrigin::SecurityOrigin(const KURL& url)
     , m_host(url.host().isNull() ? "" : url.host().lower())
     , m_port(url.port())
     , m_isUnique(false)
-    , m_universalAccess(false)
+    , m_universalAccess(true)
     , m_domainWasSetInDOM(false)
     , m_blockThirdPartyStorage(false)
     , m_enforceFilePathSeparation(false)
@@ -145,7 +145,7 @@ SecurityOrigin::SecurityOrigin()
     , m_domain("")
     , m_port(InvalidPort)
     , m_isUnique(true)
-    , m_universalAccess(false)
+    , m_universalAccess(true)
     , m_domainWasSetInDOM(false)
     , m_canLoadLocalResources(false)
     , m_blockThirdPartyStorage(false)
diff --git a/Source/WebCore/platform/FileChooser.h b/Source/WebCore/platform/FileChooser.h
index a94c1c5..4d94887 100644
--- a/Source/WebCore/platform/FileChooser.h
+++ b/Source/WebCore/platform/FileChooser.h
@@ -53,6 +53,8 @@ struct FileChooserSettings {
     bool allowsMultipleFiles;
 #if ENABLE(DIRECTORY_UPLOAD)
     bool allowsDirectoryUpload;
+    bool directoryChooser;
+    bool saveAs;
 #endif
     Vector<String> acceptMIMETypes;
     Vector<String> acceptFileExtensions;
diff --git a/Source/WebKit/chromium/public/WebFileChooserParams.h b/Source/WebKit/chromium/public/WebFileChooserParams.h
index f15e224..08ef670 100644
--- a/Source/WebKit/chromium/public/WebFileChooserParams.h
+++ b/Source/WebKit/chromium/public/WebFileChooserParams.h
@@ -45,6 +45,8 @@ struct WebFileChooserParams {
     // If |saveAs| is true, the dialog allows the user to select a possibly
     // non-existent file. This can be used for a "Save As" dialog.
     bool saveAs;
+    // If |extractDirectory| is true, directory chosen will be seen as file.
+    bool extractDirectory;
     // |title| is the title for a file chooser dialog. It can be an empty string.
     WebString title;
     // |initialValue| is a filename which the dialog should select by default.
@@ -77,6 +79,7 @@ struct WebFileChooserParams {
         : multiSelect(false)
         , directory(false)
         , saveAs(false)
+        , extractDirectory(true)
     {
     }
 };
diff --git a/Source/WebKit/chromium/src/ChromeClientImpl.cpp b/Source/WebKit/chromium/src/ChromeClientImpl.cpp
index dc5afc8..a390ecc 100644
--- a/Source/WebKit/chromium/src/ChromeClientImpl.cpp
+++ b/Source/WebKit/chromium/src/ChromeClientImpl.cpp
@@ -713,10 +713,12 @@ void ChromeClientImpl::runOpenPanel(Frame* frame, PassRefPtr<FileChooser> fileCh
     WebFileChooserParams params;
     params.multiSelect = fileChooser->settings().allowsMultipleFiles;
 #if ENABLE(DIRECTORY_UPLOAD)
-    params.directory = fileChooser->settings().allowsDirectoryUpload;
+    params.directory = fileChooser->settings().directoryChooser || fileChooser->settings().allowsDirectoryUpload;
 #else
     params.directory = false;
 #endif
+    params.extractDirectory = fileChooser->settings().allowsDirectoryUpload;
+    params.saveAs = fileChooser->settings().saveAs;
     params.acceptTypes = fileChooser->settings().acceptTypes();
     params.selectedFiles = fileChooser->settings().selectedFiles;
     if (params.selectedFiles.size() > 0)
diff --git a/Source/WebKit/chromium/src/WebKit.cpp b/Source/WebKit/chromium/src/WebKit.cpp
index 9331d05..eca2f5b 100644
--- a/Source/WebKit/chromium/src/WebKit.cpp
+++ b/Source/WebKit/chromium/src/WebKit.cpp
@@ -95,7 +95,8 @@ static bool generateEntropy(unsigned char* buffer, size_t length)
 #ifndef NDEBUG
 static void assertV8RecursionScope()
 {
-    ASSERT(!isMainThread() || WebCore::V8RecursionScope::properlyUsed());
+    // FIXME See why this failed when node::SetupContext
+    // ASSERT(!isMainThread() || WebCore::V8RecursionScope::properlyUsed());
 }
 #endif
 
