diff --git a/src/libANGLE/renderer/d3d/d3d11/win32/NativeWindow.cpp b/src/libANGLE/renderer/d3d/d3d11/win32/NativeWindow.cpp
index 05d7a46..bb0607e 100644
--- src/libANGLE/renderer/d3d/d3d11/win32/NativeWindow.cpp
+++ src/libANGLE/renderer/d3d/d3d11/win32/NativeWindow.cpp
@@ -139,6 +139,7 @@ HRESULT NativeWindow::createSwapChain(ID3D11Device* device, DXGIFactory* factory
         }
         mVisual->SetContent(swapChain1);
         mCompositionTarget->SetRoot(mVisual);
+        result = factory2->MakeWindowAssociation(mWindow, DXGI_MWA_NO_ALT_ENTER);
         SafeRelease(factory2);
         return result;
     }
@@ -167,6 +168,7 @@ HRESULT NativeWindow::createSwapChain(ID3D11Device* device, DXGIFactory* factory
         {
             *swapChain = static_cast<DXGISwapChain*>(swapChain1);
         }
+        result = factory2->MakeWindowAssociation(mWindow, DXGI_MWA_NO_ALT_ENTER);
         SafeRelease(factory2);
         return result;
     }
@@ -189,7 +191,11 @@ HRESULT NativeWindow::createSwapChain(ID3D11Device* device, DXGIFactory* factory
     swapChainDesc.Windowed = TRUE;
     swapChainDesc.SwapEffect = DXGI_SWAP_EFFECT_DISCARD;
 
-    return factory->CreateSwapChain(device, &swapChainDesc, swapChain);
+    HRESULT result = factory->CreateSwapChain(device, &swapChainDesc, swapChain);
+    if (FAILED(result)) {
+        return result;
+    }
+    return factory->MakeWindowAssociation(mWindow, DXGI_MWA_NO_ALT_ENTER);
 }
 
 void NativeWindow::commitChange()
