diff --git a/third_party/libc++/BUILD.gn b/third_party/libc++/BUILD.gn
index 9bcd072..bb61518 100644
--- third_party/libc++/BUILD.gn
+++ third_party/libc++/BUILD.gn
@@ -21,7 +21,11 @@ config("config") {
 if (libcxx_is_shared) {
   _libcxx_target_type = "shared_library"
 } else {
-  _libcxx_target_type = "source_set"
+  if (is_linux) {
+    _libcxx_target_type = "static_library"
+  } else {
+    _libcxx_target_type = "source_set"
+  }
 }
 target(_libcxx_target_type, "libc++") {
   if (libcxx_is_shared) {
diff --git a/third_party/libc++abi/BUILD.gn b/third_party/libc++abi/BUILD.gn
index 759327c..7df1531 100644
--- third_party/libc++abi/BUILD.gn
+++ third_party/libc++abi/BUILD.gn
@@ -4,7 +4,7 @@
 
 import("//build/config/c++/c++.gni")
 
-source_set("libc++abi") {
+static_library("libc++abi") {
   visibility = [
     "//buildtools/third_party/libc++",
     "//build/config:executable_deps",
