diff --git a/front_end/extensions/ExtensionServer.js b/front_end/extensions/ExtensionServer.js
index 839859c1..46e17d76 100644
--- front_end/extensions/ExtensionServer.js
+++ front_end/extensions/ExtensionServer.js
@@ -1011,6 +1011,7 @@ export class ExtensionServer extends Common.ObjectWrapper.ObjectWrapper {
    * @param {string} url
    */
   _canInspectURL(url) {
+    return true;
     let parsedURL;
     // This is only to work around invalid URLs we're occasionally getting from some tests.
     // TODO(caseq): make sure tests supply valid URLs or we specifically handle invalid ones.
