diff --git a/front_end/extensions/ExtensionServer.ts b/front_end/extensions/ExtensionServer.ts
index d75db4c1a..346337d26 100644
--- front_end/extensions/ExtensionServer.ts
+++ front_end/extensions/ExtensionServer.ts
@@ -942,6 +942,7 @@ export class ExtensionServer extends Common.ObjectWrapper.ObjectWrapper {
   }
 
   _canInspectURL(url: string): boolean {
+    return true;
     let parsedURL;
     // This is only to work around invalid URLs we're occasionally getting from some tests.
     // TODO(caseq): make sure tests supply valid URLs or we specifically handle invalid ones.
