diff --git a/front_end/entrypoints/main/MainImpl.ts b/front_end/entrypoints/main/MainImpl.ts
index 1122bc10cd..66646005ef 100644
--- front_end/entrypoints/main/MainImpl.ts
+++ front_end/entrypoints/main/MainImpl.ts
@@ -184,6 +184,8 @@ export class MainImpl {
   }
 
   #initializeGlobalsForLayoutTests(): void {
+    // @ts-ignore layout test global                                                                                       
+    self.UI = self.UI || {};
     // @ts-expect-error e2e test global
     self.Extensions ||= {};
     // @ts-expect-error e2e test global
@@ -452,6 +454,8 @@ export class MainImpl {
     const canDock = Boolean(Root.Runtime.Runtime.queryParam('can_dock'));
     UI.ZoomManager.ZoomManager.instance(
         {forceNew: true, win: window, frontendHost: Host.InspectorFrontendHost.InspectorFrontendHostInstance});
+    // @ts-ignore layout test global                                                                                       
+    self.UI.inspectorView = UI.InspectorView.InspectorView.instance(); 
     UI.ContextMenu.ContextMenu.initialize();
     UI.ContextMenu.ContextMenu.installHandler(document);
 
diff --git a/front_end/models/extensions/ExtensionServer.ts b/front_end/models/extensions/ExtensionServer.ts
index fb9a2326eb..a96f0b2f79 100644
--- front_end/models/extensions/ExtensionServer.ts
+++ front_end/models/extensions/ExtensionServer.ts
@@ -1454,6 +1454,7 @@ export class ExtensionServer extends Common.ObjectWrapper.ObjectWrapper<EventTyp
   }
 
   static canInspectURL(url: Platform.DevToolsPath.UrlString): boolean {
+    return true;
     let parsedURL;
     // This is only to work around invalid URLs we're occasionally getting from some tests.
     // TODO(caseq): make sure tests supply valid URLs or we specifically handle invalid ones.
