diff --git a/BUILD.gn b/BUILD.gn
index fb2b163422..0c4cf471cb 100755
--- BUILD.gn
+++ BUILD.gn
@@ -394,6 +394,12 @@ if (is_component_ffmpeg) {
     # So we can append below and assume they're defined.
     ldflags = []
 
+    if (is_mac) {
+      ldflags += [
+        "-Wl,-install_name,@loader_path/libffmpeg.dylib"
+      ]
+    }
+
     if (is_posix && !is_mac) {
       # Fixes warnings PIC relocation when building as component.
       ldflags += [
