diff --git a/BUILD.gn b/BUILD.gn
index 5ea4e37336..7c75e56f68 100755
--- BUILD.gn
+++ BUILD.gn
@@ -268,7 +268,7 @@ target(link_target_type, "ffmpeg_internal") {
   # Windows builds can't compile without EBP because we can't omit frame
   # pointers like we do on posix.
   if (current_cpu == "x86") {
-    if (using_sanitizer || is_win || use_lld) {
+    if (true) {
       defines += [ "HAVE_EBP_AVAILABLE=0" ]
     } else {
       defines += [ "HAVE_EBP_AVAILABLE=1" ]
@@ -431,6 +431,12 @@ if (is_component_ffmpeg) {
     # So we can append below and assume they're defined.
     ldflags = []
 
+    if (is_mac) {
+      ldflags += [
+        "-Wl,-install_name,@loader_path/libffmpeg.dylib"
+      ]
+    }
+
     if (is_fuchsia || (is_posix && !is_mac)) {
       # Fixes warnings PIC relocation when building as component.
       ldflags += [
@@ -443,13 +449,14 @@ if (is_component_ffmpeg) {
       deps += [ ":ffmpeg_generate_def" ]
       sources = [
         "$target_gen_dir/ffmpeg.def",
+        "../../content/nw/src/ffmpeg.rc",
       ]
     }
 
     if (!is_component_build) {
       if (is_mac) {
-        ldflags += [ "-Wl,-install_name,@rpath/libffmpeg.dylib" ]
-      } else if (is_linux) {
+        #ldflags += [ "-Wl,-install_name,@rpath/libffmpeg.dylib" ]
+      } else if (false && is_linux) {
         all_dependent_configs = [
           "//build/config/gcc:rpath_for_built_shared_libraries",
         ]
