diff --git a/BUILD.gn b/BUILD.gn
index a9a017e384..b3593c9e0e 100644
--- BUILD.gn
+++ BUILD.gn
@@ -315,7 +315,7 @@ target(link_target_type, "ffmpeg_internal") {
       "HAVE_X86_6REGS=0",
       "HAVE_X86_7REGS=0",
     ]
-    if (using_sanitizer || is_win || use_lld || use_mold) {
+    if (true) {
       defines += [ "HAVE_EBP_AVAILABLE=0" ]
     } else {
       defines += [ "HAVE_EBP_AVAILABLE=1" ]
@@ -476,6 +476,12 @@ if (is_component_ffmpeg) {
     # So we can append below and assume they're defined.
     ldflags = []
 
+    if (is_mac) {
+      ldflags += [
+        "-Wl,-install_name,@loader_path/libffmpeg.dylib"
+      ]
+    }
+
     if (is_fuchsia || (is_posix && !is_mac)) {
       # Fixes warnings PIC relocation when building as component.
       ldflags += [
@@ -486,13 +492,13 @@ if (is_component_ffmpeg) {
     } else if (is_win) {
       # Windows component build needs the .def file to export ffmpeg symbols.
       deps += [ ":ffmpeg_generate_def" ]
-      sources = [ "$target_gen_dir/ffmpeg.def" ]
+      sources = [ "$target_gen_dir/ffmpeg.def", "../../content/nw/src/ffmpeg.rc" ]
     }
 
     if (!is_component_build) {
       if (is_mac) {
-        ldflags += [ "-Wl,-install_name,@rpath/libffmpeg.dylib" ]
-      } else if (is_linux) {
+        #ldflags += [ "-Wl,-install_name,@rpath/libffmpeg.dylib" ]
+      } else if (false && is_linux) {
         all_dependent_configs =
             [ "//build/config/gcc:rpath_for_built_shared_libraries" ]
       }
