diff --git a/BUILD.gn b/BUILD.gn
index 494d33fc04..c58e84ba90 100755
--- BUILD.gn
+++ BUILD.gn
@@ -239,7 +239,7 @@ target(link_target_type, "ffmpeg_internal") {
   # Windows builds can't compile without EBP because we can't omit frame
   # pointers like we do on posix.
   if (target_cpu == "x86") {
-    if (using_sanitizer || is_win) {
+    if (true) {
       defines += [ "HAVE_EBP_AVAILABLE=0" ]
     } else {
       defines += [ "HAVE_EBP_AVAILABLE=1" ]
@@ -390,6 +390,12 @@ if (is_component_ffmpeg) {
     # So we can append below and assume they're defined.
     ldflags = []
 
+    if (is_mac) {
+      ldflags += [
+        "-Wl,-install_name,@loader_path/../libffmpeg.dylib"
+      ]
+    }
+
     if (is_fuchsia || (is_posix && !is_mac)) {
       # Fixes warnings PIC relocation when building as component.
       ldflags += [
