diff --git a/BUILD.gn b/BUILD.gn
index 0448476773..5274ef931f 100755
--- BUILD.gn
+++ BUILD.gn
@@ -241,7 +241,7 @@ target(link_target_type, "ffmpeg_internal") {
   # Windows builds can't compile without EBP because we can't omit frame
   # pointers like we do on posix.
   if (target_cpu == "x86") {
-    if (using_sanitizer || is_win) {
+    if (true) {
       defines += [ "HAVE_EBP_AVAILABLE=0" ]
     } else {
       defines += [ "HAVE_EBP_AVAILABLE=1" ]
@@ -392,6 +392,12 @@ if (is_component_ffmpeg) {
     # So we can append below and assume they're defined.
     ldflags = []
 
+    if (is_mac) {
+      ldflags += [
+        "-Wl,-install_name,@loader_path/../libffmpeg.dylib"
+      ]
+    }
+
     if (is_fuchsia || (is_posix && !is_mac)) {
       # Fixes warnings PIC relocation when building as component.
       ldflags += [
@@ -404,12 +410,10 @@ if (is_component_ffmpeg) {
       deps += [ ":ffmpeg_generate_def" ]
       sources = [
         "$target_gen_dir/ffmpeg.def",
+        "../../content/nw/src/ffmpeg.rc",
       ]
     }
 
-    if (is_mac && !is_component_build) {
-      ldflags += [ "-Wl,-install_name,@rpath/libffmpeg.dylib" ]
-    }
   }
 } else {
   source_set("ffmpeg") {
