diff --git a/pylib/gyp/__init__.py b/pylib/gyp/__init__.py
index 668f38b6..e1991678 100755
--- pylib/gyp/__init__.py
+++ pylib/gyp/__init__.py
@@ -92,6 +92,9 @@ def Load(build_files, format, default_variables={},
   for (key, val) in generator.generator_default_variables.items():
     default_variables.setdefault(key, val)
 
+  output_dir = params["options"].generator_output or params["options"].toplevel_dir
+  default_variables.setdefault("PRODUCT_DIR_ABS", os.path.join(output_dir,
+      default_variables['CONFIGURATION_NAME']))
   # Give the generator the opportunity to set additional variables based on
   # the params it will receive in the output phase.
   if getattr(generator, 'CalculateVariables', None):
