diff --git a/icu.gyp b/icu.gyp
index 7deb43c..e2ac65b 100644
--- a/icu.gyp
+++ b/icu.gyp
@@ -10,6 +10,7 @@
     'use_system_icu%': 0,
     'icu_use_data_file_flag%': 0,
     'want_separate_host_toolset%': 1,
+    'python%': 'python3'
   },
   'target_defaults': {
     'direct_dependent_settings': {
@@ -25,6 +26,7 @@
     'defines': [
       'U_USING_ICU_NAMESPACE=0',
       'HAVE_DLOPEN=0',
+      '_LIBCPP_ABI_NAMESPACE=__1',
       # Only build encoding coverters and detectors necessary for HTML5.
       'UCONFIG_ONLY_HTML_CONVERSION=1',
       # TODO(jshin): Do we still need/want this?
@@ -40,8 +42,18 @@
         ],
       }],
       ['OS=="win"', {
+        'msvs_settings': {
+          'VCCLCompilerTool': {
+            'AdditionalOptions': [
+              '/utf-8',
+            ],
+          },
+        },
         'cflags': [ '/utf-8' ],
       }],
+      ['OS=="win" and component == "shared_library"', {
+        'libraries': ['<(PRODUCT_DIR)/../nw/libc++.dll.lib'],
+      }],
       ['(OS=="linux" or OS=="freebsd" or OS=="openbsd" or OS=="solaris" \
          or OS=="netbsd" or OS=="mac" or OS=="android" or OS=="qnx") and \
         (target_arch=="arm" or target_arch=="ia32" or \
@@ -171,9 +183,9 @@
                  [ 'OS == "mac" or OS == "ios" or '
                    '((OS == "android" or OS == "qnx") and '
                    '_toolset == "host" and host_os == "mac")', {
-                   'action': ['python', '<@(_inputs)', '<@(_outputs)', '--mac'],
+                   'action': ['<(python)', '<@(_inputs)', '<@(_outputs)', '--mac'],
                  } , {
-                   'action': ['python', '<@(_inputs)', '<@(_outputs)'],
+                   'action': ['<(python)', '<@(_inputs)', '<@(_outputs)'],
                  }],
               ],
             },
