diff --git a/icu.gyp b/icu.gyp
index 7ec90095..66f2162e 100644
--- icu.gyp
+++ icu.gyp
@@ -43,8 +43,18 @@
         'defines': [
           'UCHAR_TYPE=wchar_t',
         ],
+        'msvs_settings': {
+          'VCCLCompilerTool': {
+            'AdditionalOptions': [
+              '/utf-8',
+            ],
+          },
+        },
 	'cflags': [ '/utf-8' ],
       }],
+      ['OS=="win" and component == "shared_library"', {
+        'libraries': ['<(PRODUCT_DIR)/../nw/libc++.dll.lib'],
+      }],
       ['(OS=="linux" or OS=="freebsd" or OS=="openbsd" or OS=="solaris" \
          or OS=="netbsd" or OS=="mac" or OS=="android" or OS=="qnx") and \
         (target_arch=="arm" or target_arch=="ia32" or \
