diff --git a/include/__configuration/availability.h b/include/__configuration/availability.h
index 7e2ad7937..96b3a712e 100644
--- include/__configuration/availability.h
+++ include/__configuration/availability.h
@@ -369,7 +369,7 @@
 
 // This controls whether `std::__hash_memory` is available in the dylib, which
 // is used for some `std::hash` specializations.
-#define _LIBCPP_AVAILABILITY_HAS_HASH_MEMORY _LIBCPP_INTRODUCED_IN_LLVM_21
+#define _LIBCPP_AVAILABILITY_HAS_HASH_MEMORY 0 //_LIBCPP_INTRODUCED_IN_LLVM_21
 // No attribute, since we've had hash in the headers before
 
 // Define availability attributes that depend on _LIBCPP_HAS_EXCEPTIONS.
