diff --git a/BUILD.gn b/BUILD.gn
index dbc0a546a5..dbe8210d19 100644
--- BUILD.gn
+++ BUILD.gn
@@ -278,6 +278,11 @@ if (!build_with_chromium) {
 # Rationalize them with eseckler@. For now seems this one is only used from
 # chromium and the other one only from the Android tree.
 if (build_with_chromium) {
+  static_library("libperfettonode") {
+    complete_static_lib = true
+    deps = [ ":libperfetto" ]
+  }
+
   component("libperfetto") {
     public_configs = [ "gn:public_config" ]
     deps = [
diff --git a/gn/perfetto.gni b/gn/perfetto.gni
index 4fb532064c..d545454eba 100644
--- gn/perfetto.gni
+++ gn/perfetto.gni
@@ -511,3 +511,6 @@ assert(!perfetto_use_pkgconfig || perfetto_use_system_protobuf,
 assert(
     !enable_perfetto_llvm_symbolizer || !use_custom_libcxx,
     "enable_perfetto_llvm_symbolizer and use_custom_libcxx are mutually exclusive. If you are seeing this error then you have set enable_perfetto_llvm_symbolizer=true and must set use_custom_libcxx=false to use llvm_symbolizer.")
+
+# NW.js: disable zlib in perfetto to avoid conflicts with node's zlib
+enable_perfetto_zlib = false
