diff --git a/selenium/webdriver/chromium/options.py b/selenium/webdriver/chromium/options.py
index 3d8f572..83220d8 100644
--- selenium/webdriver/chromium/options.py
+++ selenium/webdriver/chromium/options.py
@@ -31,6 +31,7 @@ class ChromiumOptions(ArgOptions):
         super(ChromiumOptions, self).__init__()
         self._binary_location = ''
         self._extension_files = []
+        self._nw_arguments = []
         self._extensions = []
         self._experimental_options = {}
         self._debugger_address = None
@@ -69,6 +70,25 @@ class ChromiumOptions(ArgOptions):
         """
         self._debugger_address = value
 
+    @property
+    def nw_arguments(self):
+        """
+        Returns a list of arguments needed for the browser
+        """
+        return self._nw_arguments
+
+    def add_nw_argument(self, argument):
+        """
+        Adds an argument to the list
+
+        :Args:
+         - Sets the arguments
+        """
+        if argument:
+            self._nw_arguments.append(argument)
+        else:
+            raise ValueError("argument can not be null")
+
     @property
     def extensions(self) -> List[str]:
         """
@@ -167,6 +187,7 @@ class ChromiumOptions(ArgOptions):
         if self.binary_location:
             chrome_options["binary"] = self.binary_location
         chrome_options["args"] = self._arguments
+        chrome_options["nwargs"] = self.nw_arguments
         if self.debugger_address:
             chrome_options["debuggerAddress"] = self.debugger_address
 
