diff --git a/layers/synchronization_validation.cpp b/layers/synchronization_validation.cpp
index f9e46c5c..6b19dff3 100644
--- layers/synchronization_validation.cpp
+++ layers/synchronization_validation.cpp
@@ -18,6 +18,7 @@
  * Author: Locke Lin <locke@lunarg.com>
  * Author: Jeremy Gebben <jeremyg@lunarg.com>
  */
+#pragma clang diagnostic ignored "-Wrange-loop-construct"
 
 #include <limits>
 #include <vector>
